File: frontend/src/types/session.ts
export interface SessionAnswer {
  id: number
  session_id: number
  question_id: number
  answer_value: string
  created_at: string
  updated_at: string
}

export interface DocumentSession {
  id: number
  client_identifier: string
  user_id: number
  current_group_id: number | null
  is_completed: boolean
  completed_at: string | null
  created_at: string
  updated_at: string
}

export interface DocumentSessionWithAnswers extends DocumentSession {
  answers: SessionAnswer[]
}

export interface SessionCreate {
  client_identifier: string
  starting_group_id?: number
}

export interface SubmitAnswersRequest {
  answers: Array<{
    question_id: number
    answer_value: string
  }>
}

export interface SessionProgress {
  session: DocumentSession
  current_group: {
    id: number
    name: string
    description: string | null
    questions: Array<{
      id: number
      identifier: string
      question_text: string
      question_type: string
      options: any
      validation_rules: any
      is_required: boolean
    }>
  } | null
  next_group_id: number | null
  is_completed: boolean
  total_answers: number
}

export interface QuestionToDisplay {
  id: number
  identifier: string
  question_text: string
  question_type: string
  is_required: boolean
  repeatable: boolean
  help_text: string | null
  options: Array<{ value: string; label: string }> | null
  person_display_mode: string | null
  include_time: boolean | null
  validation_rules: Record<string, any> | null
  current_answer: string | null
  depth: number  // Nesting level for conditional questions
}

export interface SessionQuestionsResponse {
  session_id: number
  client_identifier: string
  flow_id: number | null
  flow_name: string | null
  current_group_id: number
  current_group_name: string
  current_group_index: number
  total_groups: number
  questions: QuestionToDisplay[]
  current_page: number
  total_pages: number
  questions_per_page: number
  is_completed: boolean
  is_last_group: boolean
  can_go_back: boolean
  existing_answers: Record<number, string>
  conditional_identifiers: string[]  // identifiers that have conditionals depending on them
}

export interface SaveAnswersRequest {
  answers: Array<{
    question_id: number
    answer_value: string
  }>
}

export interface NavigateRequest {
  direction: 'forward' | 'backward'
  answers?: Array<{
    question_id: number
    answer_value: string
  }>
}

File: backend/src/schemas/session.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class SessionAnswerBase(BaseModel):
    """Base schema for session answers."""
    question_id: int
    answer_value: str


class SessionAnswerCreate(SessionAnswerBase):
    """Schema for creating a session answer."""
    pass


class SessionAnswerResponse(SessionAnswerBase):
    """Schema for session answer response."""
    id: int
    session_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class InputFormCreate(BaseModel):
    """Schema for creating an input form."""
    client_identifier: str = Field(..., min_length=1, max_length=255)
    flow_id: Optional[int] = None
    starting_group_id: Optional[int] = None


class InputFormUpdate(BaseModel):
    """Schema for updating an input form."""
    current_group_id: Optional[int] = None
    is_completed: Optional[bool] = None


class InputFormResponse(BaseModel):
    """Schema for input form response."""
    id: int
    client_identifier: str
    user_id: int
    flow_id: Optional[int]
    current_group_id: Optional[int]
    current_group_name: Optional[str] = None
    is_completed: bool
    completed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class InputFormWithAnswers(InputFormResponse):
    """Schema for session with answers."""
    answers: List[SessionAnswerResponse]


class SubmitAnswersRequest(BaseModel):
    """Schema for submitting answers to a question group."""
    answers: List[SessionAnswerCreate]


class SessionProgressResponse(BaseModel):
    """Schema for session progress response."""
    session: InputFormResponse
    current_group: Optional[dict]
    next_group_id: Optional[int]
    is_completed: bool
    total_answers: int


class QuestionToDisplay(BaseModel):
    """Schema for a question to display in the document."""
    id: int
    identifier: str
    question_text: str
    question_type: str
    is_required: bool
    repeatable: bool = False
    repeatable_group_id: Optional[str] = None
    help_text: Optional[str]
    options: Optional[List[dict]]
    person_display_mode: Optional[str]
    include_time: Optional[bool]
    validation_rules: Optional[dict]
    current_answer: Optional[str] = None
    depth: int = 0  # Nesting level for conditional questions


class SessionQuestionsResponse(BaseModel):
    """Schema for session questions response with pagination."""
    session_id: int
    client_identifier: str
    flow_id: Optional[int]
    flow_name: Optional[str]
    current_group_id: int
    current_group_name: str
    current_group_index: int
    total_groups: int
    questions: List[QuestionToDisplay]
    current_page: int
    total_pages: int
    questions_per_page: int
    is_completed: bool
    is_last_group: bool
    can_go_back: bool
    existing_answers: dict  # question_id -> answer_value
    conditional_identifiers: List[str] = []  # identifiers that have conditionals depending on them


class SaveAnswersRequest(BaseModel):
    """Schema for saving answers (without navigating)."""
    answers: List[SessionAnswerCreate]


class NavigateRequest(BaseModel):
    """Schema for navigation request."""
    direction: str = Field(..., pattern="^(forward|backward)$")
    answers: Optional[List[SessionAnswerCreate]] = None
